modules/Members/classes/MemberListProvider.php

```diff
@@ -0,0 +1,179 @@
+        return URL::build('/members', 'list=' . $this->getName());
+    }
+
+    /**
+     * Determine whether the member list provider is enabled or not. Will automatically enable the member list if it is not already enabled.
+     * @return bool Whether the member list provider is enabled
+     */
+    public function isEnabled(): bool {
+        if (isset($this->_enabled)) {
+            return $this->_enabled;
+        }
+
+        $enabled = DB::getInstance()->get('member_lists', ['name', $this->getName()])->first()->enabled;
+        if ($enabled === null) {
+            DB::getInstance()->insert('member_lists', [
+                'name' => $this->getName(),
+                'friendly_name' => $this->getFriendlyName(),
+                'module' => $this->getModule(),
+                'enabled' => true,
+            ]);
+
+            return true;
+        }
+
+        return $this->_enabled = $enabled;
+    }
+
+    /**
+     * Get the information needed to generate the member list.
+     *
+     * @return array An array containing the SQL query to run, the column name of the user ID, and optionally the
+     * column name of the "count" value for this list. Count values are used to display the number of posts, likes, etc.
+     */
+    abstract protected function generator(): array;
+
+    /**
+     * Get an array of members to display on the member list page.
+     *
+     * @param bool $overview Whether the member list is being displayed on the overview page. If true, only 5 members will be returned, otherwise 20.
+     * @param int $page The page number to display, starting at 1 - pages are 20 members long
+     * @return array An array of members to display on the member list page
+     */
+    public function getMembers(bool $overview, int $page): array {
+        [$sql, $id_column, $count_column] = $this->generator();
+
+        $rows = DB::getInstance()->query($sql)->results();
+        if (Util::getSetting('member_list_hide_banned', false, 'Members')) {
+            $rows = $this->filterBanned($rows, $id_column);
+        }
+
+        $rows = array_slice($rows, ($page - 1) * 20);
+
+        $list_members = [];
+        $limit = $overview ? 5 : 20;
+
+
+        foreach ($rows as $row) {
+            if (count($list_members) === $limit) {
+                break;
+            }
+
+            $user_id = $row->{$id_column};
+            $member = new User($user_id);
+
+            $list_members[] = array_merge(
+                [
+                    'username' => $member->data()->username,
+                    'avatar_url' => $member->getAvatar(),
+                    'group_style' => $member->getGroupStyle(),
+                    'profile_url' => $member->getProfileURL(),
+                    'count' => $count_column ? $row->{$count_column} : null,
+                ],
+                $overview
+                    ? []
+                    : [
+                        'group_html' => implode('', $member->getAllGroupHtml()),
```

Comment:
Assuming this gets displayed as a list of all users' groups, should this be passed through to the template as an array so the template can separate the group tags however it wants

Reply:
Great idea